vpr/src/timing/timing_util.cpp

```diff
@@ -637,6 +643,21 @@ void print_hold_timing_summary(const tatum::TimingConstraints& constraints, cons
         }
     }
     VTR_LOG("\n");
+
+    check_if_failed_timing_constraints(hold_worst_neg_slack, "hold Worst");
+    check_if_failed_timing_constraints(hold_total_neg_slack, "hold Total");
+}
+
+/* Checks if slack is negative and if user wants VPR flow to fail if their desisn doesn't meet timing
+ * constraints. If both conditions are true, the function adds details about the negative slack to a string
+ * that will be printed when VPR throws an error.
+ */
+void check_if_failed_timing_constraints(double& slack, std::string slack_name) {
+    auto& timing_ctx = g_vpr_ctx.mutable_timing();
+    if (timing_ctx.fail_status_if_fail_timing && slack < 0) {
+        timing_ctx.fail_timing_msg = timing_ctx.fail_timing_msg + "Final " + std::to_string(slack) + " Negative Slack of " + slack_name.c_str() + " was found to be negative. \n";
```

Comment:
avoid using fail_timing_msg (passing a variable all the way through a top level context to a single routine -- set up the message here instead).
Message should be more explicit: say you're terminating processing due to a negative slack and option xx set.